src/normal.cc

```diff
@@ -88,11 +88,17 @@ template<InsertMode mode>
 void enter_insert_mode(Context& context, NormalParams)
 {
     context.input_handler().insert(mode);
+
```

Comment:
I dont like all this repetition in every function, not sure what the correct solution would be (maybe Buffer should throw a runtime_error when we try to insert in a read only buffer), but this change is too intrusive.